plugins/out_s3/s3.c

```diff
@@ -980,28 +1046,51 @@ static int s3_put_object(struct flb_s3 *ctx, const char *tag, time_t create_time
     flb_sds_destroy(s3_key);
     uri = tmp;
 
+    if (ctx->use_put_object == FLB_TRUE && ctx->compression != NULL) {
+        if (strcmp(ctx->compression, "gzip") != 0) {
+            flb_plg_error(ctx->ins, 
+                          "'gzip' is the only support value of compression field for now");
+            flb_sds_destroy(uri);
+            return -1;
+        }
+        ret = flb_gzip_compress(body, body_size, &final_body, &final_body_size);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "Failed to compress data");
+            flb_sds_destroy(uri);
+            return -1;
+        }
+    } else {
+        final_body = body;
+        final_body_size = body_size;
+    }
+    if (ctx->content_type != NULL) {
+        num_headers++;
+    }
+    if (ctx->compression != NULL) {
+        num_headers++;
+    }
+    if (ctx->canned_acl != NULL) {
+        num_headers++;
+    }
+    
     s3_client = ctx->s3_client;
     if (s3_plugin_under_test() == FLB_TRUE) {
         c = mock_s3_call("TEST_PUT_OBJECT_ERROR", "PutObject");
     }
     else {
-        if (ctx->canned_acl == NULL) {
-            c = s3_client->client_vtable->request(s3_client, FLB_HTTP_PUT,
-                                                  uri, body, body_size,
-                                                  NULL, 0);
+        headers = create_headers(ctx, num_headers);
+        if (headers == NULL && num_headers != 0) {
+            flb_plg_error(ctx->ins, "Failed to create headers");
+            flb_sds_destroy(uri);
+            return -1;
         }
-        else {
-            canned_acl_header = create_canned_acl_header(ctx->canned_acl);
-            if (canned_acl_header == NULL) {
-                flb_sds_destroy(uri);
-                flb_plg_error(ctx->ins, "Failed to create canned ACL header");
-                return -1;
-            }
-            c = s3_client->client_vtable->request(s3_client, FLB_HTTP_PUT,
-                                                  uri, body, body_size,
-                                                  canned_acl_header, 1);
-            flb_free(canned_acl_header);
+        c = s3_client->client_vtable->request(s3_client, FLB_HTTP_PUT,
+                                              uri, final_body, final_body_size,
+                                              headers, num_headers);
+        if (ctx->canned_acl != NULL) {
+            flb_free(&headers[num_headers-1]);
```

Comment:
Actually I am thinking about if it is a good way to free canned_acl_header. But I haven't thought about a better and more elegant way yet..

Reply:
I am confused, why not free the entire header array- with:

```
flb_free(headers);
```

The headers array is a single block of memory

Reply:
I checked online and it says: "You will need to free what you have allocated."  I free the entire header array and the canned_acl_header is also allocated by create_acl_header. So I want to free that one as well..

Reply:
yeah see my other comments, you should refactor so that there is only one allocation.

(Also even if you allocate the canned header separately, this will not free it. If that doesn't make sense, call me and we can draw out what the memory actually looks like (or try that yourself)).

Reply:
Thanks. I would try by myself first to see if I can understand it.